cassiopeia/solver/distance_correction/_crispr_cas9_distance_correction_solver.py

```diff
@@ -0,0 +1,556 @@
+"""
+This file stores the CRISPRCas9DistanceCorrectionSolver class, which enables the
+estimation of tree topologies from _corrected_ distances, which are typically
+superior to raw distances such as the Hamming distance.
+"""
+from functools import partial
+from typing import Callable, Dict, List, Optional
+
+import numpy as np
+import pandas as pd
+
+from cassiopeia.data import CassiopeiaTree
+
+from ..CassiopeiaSolver import CassiopeiaSolver
+from ..DistanceSolver import DistanceSolver
+from ..NeighborJoiningSolver import NeighborJoiningSolver
+
+CharacterMatrixType = pd.DataFrame
+CharacterStatesType = List[int]
+MissingDataIndicatorType = int
+MutationProportionType = float
+CollisionProbabilityType = float
+MutationProportionEstimatorType = Callable[
+    [CharacterMatrixType], MutationProportionType
+]
+CollisionProbabilityEstimatorType = Callable[
+    [CharacterMatrixType], CollisionProbabilityType
+]
+CRISPRCas9DistanceCorrectorType = Callable[
+    [
+        MutationProportionType,
+        CollisionProbabilityType,
+        CharacterStatesType,
+        CharacterStatesType,
+        MissingDataIndicatorType,
+    ],
+    float,
+]
+
+
+def _inverse(
+    f: Callable[[float], float],
+    y: float,
+    lower: float,
+    upper: float,
+) -> float:
+    """
+    Invert an increasing function.
+
+    Finds x s.t. f(x) = y. Search for x is limited to the range [lower, upper].
+    Uses 30 fixed iterations of binary search.
+
+    Args:
+        f: Function to invert.
+        y: Value to invert.
+        lower: Lower bound on the inverse.
+        upper: Upper bound on the inverse.
+
+    Returns:
+        f^{-1}(y) restricted to the range [lower, upper].
+    """
+    for iteration in range(30):
+        mid = (upper + lower) / 2.0
+        if f(mid) < y:
+            lower, upper = mid, upper
+        else:
+            lower, upper = lower, mid
+    return (upper + lower) / 2.0
+
+
+def _hamming_distance(
```

Comment:
Although there is some duplication, I think in terms of readability it is better to have the simple versions of Hamming distance and ternary Hamming distance here in the same file as the solver. For example, `weighted_hamming_distance` from `cas.solver.dissimilarity_functions` does more than just the ternary case, allowing for aribtrary `weights`, which are not needed in this case; the `hamming_distance` in `cas.solver.dissimilarity_functions` has `ignore_missing_state` which is not used (and does not normalize). I agree we could extend the `hamming_distance` in `cas.solver.dissimilarity_functions` to normalize, but again I think it hampers readability. For such small components that are private to the module, I don't think there's really an issue in code duplication if it improves readability.

Of course, if in the future we want to use `cas.solver.dissimilarity_functions` instead, we can do it easily since this falls into the implementation detail category.